### PR TITLE
Add boolean & number type support to entity editor

### PR DESCRIPTION
**Pull request by yusufkinatas**
## 🌟 What is the purpose of this PR?

This PR adds functionality for viewing/editing value cells with data type of "Number" and "Boolean" at entity editor.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203150072758509/1203301695650120/f) _(internal)_
- [Design](https://app.asana.com/0/1203150072758509/1203301695650120/f) _(internal)_

## 🔍 What does this change?

- Add `No value` state to value cells
- Move `isValueEmpty` to a separate file
- Add draw & edit support for boolean & number values

## Next Steps

- Match the design of boolean selector 1-1 with design (this PR does not modify the default `Select` style we have in the project)

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Open entity editor with boolean/number types
1.  Confirm that editing/viewing boolean/number types work & look like expected

### Boolean view
Screenshot 2022-11-09 at 13 57 28

### Boolean editor
Screenshot 2022-11-09 at 13 57 33

### Number editor

Screenshot 2022-11-09 at 13 57 39

### CODE REVIEW

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/number-or-string-editor.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  return (
+    <TextField
+      sx={{ my: "1px" }}
+      autoFocus
+      value={value}
+      type={isNumber ? "number" : "text"}
+      inputMode={isNumber ? "numeric" : "text"}
+      onChange={({ target }) => {
+        const newCell = cloneDeep(cell);
+        const newValue = isNumber ? Number(target.value) : target.value;
```
Comment (CiaranMn) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/number-or-string-editor.tsx:21]: This makes it impossible to use the keyboard to delete the last number in the cell.

**Review by CiaranMn — CHANGES_REQUESTED** [review]
Looking good @yusufkinatas – spotted one thing on the input to fix

We'll also want to think at some point about how we display large numbers when the user isn't editing them - this isn't the most user-friendly (cc @nonparibus)
Screenshot 2022-11-10 at 09 56 41

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell.tsx [hunk 2]
```diff
@@ -19,8 +25,26 @@ export const renderValueCell: CustomRenderer<ValueCell> = {
...
+    if (isValueEmpty(value)) {
+      ctx.fillStyle = "#91A5BA";
+      ctx.font = "italic 14px Inter";
+      return ctx.fillText("No value", left, yCenter);
+    }
+
+    if (typeof value === "boolean") {
+      return drawTextWithIcon({
+        args,
+        text: capitalize(String(value)),
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell.tsx:39]: This again feels too clever to me – can we just do 

```suggestion
        text: value ? 'True' : 'False',
```

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  onFinishedEditing,
+}) => {
+  const { value } = cell.data.property;
+
+  return (
+    <TextField
+      sx={{ my: "1px" }}
+      select
+      SelectProps={{ defaultOpen: true }}
+      value={Number(value)}
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx:21]: `MenuItem.value` expects `number | string`, thought 0-1 makes sense for true-false

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      onChange={({ target }) => {
+        const newCell = cloneDeep(cell);
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx:23]: Nope, is it better to use in this case? What's the difference? Afaik `immer` is used to produce a new object from an existing object without modifying the original object. Doesn't `cloneDeep`  does the same thing to cell on this case?

**Comment by yusufkinatas** [comment]
Thanks for the reviews! Addressed to all the comments.

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (nathggns) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx:24]: `cloneDeep` clones the entire object – immer just clones the paths that have changed. It'll generally perform better over time. Just not sure why we don't use it.

[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx:24]: I have no reason for not to use it, just wasn't aware of the difference. Sounds reasonable!

[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-id].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/boolean-editor.tsx]: Updated here https://github.com/hashintel/hash/pull/1369/commits/54a3337b3c037625c940eadedb27b0ec41554bac

**Comment by nathggns** [comment]
Reviewing again

**Review by nathggns — APPROVED** [review]

**Pull request merged**